File: src-tauri/src/commands/metadata.rs
use crate::models::*;
use crate::services::metadata_service;
use crate::state::AppState;
use tauri::State;

#[tauri::command]
pub async fn list_tables(
    state: State<'_, AppState>,
    id: i64,
    schema: Option<String>,
    database: Option<String>,
) -> Result<Vec<TableInfo>, String> {
    metadata_service::list_tables(&state, id, schema, database)
        .await
        .map_err(String::from)
}

#[tauri::command]
pub async fn list_routines(
    state: State<'_, AppState>,
    id: i64,
    schema: Option<String>,
    database: Option<String>,
) -> Result<Vec<RoutineInfo>, String> {
    metadata_service::list_routines(&state, id, schema, database)
        .await
        .map_err(String::from)
}

#[tauri::command]
pub async fn list_events(
    state: State<'_, AppState>,
    id: i64,
    schema: Option<String>,
    database: Option<String>,
) -> Result<Vec<EventInfo>, String> {
    metadata_service::list_events(&state, id, schema, database)
        .await
        .map_err(String::from)
}

#[tauri::command]
pub async fn list_sequences(
    state: State<'_, AppState>,
    id: i64,
    schema: Option<String>,
    database: Option<String>,
) -> Result<Vec<SequenceInfo>, String> {
    metadata_service::list_sequences(&state, id, schema, database)
        .await
        .map_err(String::from)
}

#[tauri::command]
pub async fn list_types(
    state: State<'_, AppState>,
    id: i64,
    schema: Option<String>,
    database: Option<String>,
) -> Result<Vec<TypeInfo>, String> {
    metadata_service::list_types(&state, id, schema, database)
        .await
        .map_err(String::from)
}

#[tauri::command]
pub async fn list_synonyms(
    state: State<'_, AppState>,
    id: i64,
    schema: Option<String>,
    database: Option<String>,
) -> Result<Vec<SynonymInfo>, String> {
    metadata_service::list_synonyms(&state, id, schema, database)
        .await
        .map_err(String::from)
}

#[tauri::command]
pub async fn list_packages(
    state: State<'_, AppState>,
    id: i64,
    schema: Option<String>,
    database: Option<String>,
) -> Result<Vec<PackageInfo>, String> {
    metadata_service::list_packages(&state, id, schema, database)
        .await
        .map_err(String::from)
}

#[tauri::command]
pub async fn get_table_structure(
    state: State<'_, AppState>,
    id: i64,
    schema: String,
    table: String,
    database: Option<String>,
) -> Result<TableStructure, String> {
    metadata_service::get_table_structure(&state, id, schema, table, database)
        .await
        .map_err(String::from)
}

#[tauri::command]
pub async fn get_table_ddl(
    state: State<'_, AppState>,
    id: i64,
    schema: String,
    table: String,
    database: Option<String>,
) -> Result<String, String> {
    metadata_service::get_table_ddl(&state, id, schema, table, database)
        .await
        .map_err(String::from)
}

#[tauri::command]
pub async fn get_routine_ddl(
    state: State<'_, AppState>,
    id: i64,
    schema: String,
    routine_name: String,
    routine_type: String,
    database: Option<String>,
) -> Result<String, String> {
    metadata_service::get_routine_ddl(&state, id, schema, routine_name, routine_type, database)
        .await
        .map_err(String::from)
}

#[tauri::command]
pub async fn get_table_metadata(
    state: State<'_, AppState>,
    id: i64,
    schema: String,
    table: String,
    database: Option<String>,
) -> Result<TableMetadata, String> {
    metadata_service::get_table_metadata(&state, id, schema, table, database)
        .await
        .map_err(String::from)
}

#[tauri::command]
pub async fn get_schema_overview(
    state: State<'_, AppState>,
    id: i64,
    schema: Option<String>,
    database: Option<String>,
) -> Result<SchemaOverview, String> {
    metadata_service::get_schema_overview(&state, id, schema, database)
        .await
        .map_err(String::from)
}

#[tauri::command]
pub async fn get_schema_foreign_keys(
    state: State<'_, AppState>,
    id: i64,
    schema: Option<String>,
    database: Option<String>,
) -> Result<Vec<SchemaForeignKey>, String> {
    metadata_service::get_schema_foreign_keys(&state, id, schema, database)
        .await
        .map_err(String::from)
}

#[tauri::command]
pub async fn get_driver_capabilities(
    state: State<'_, AppState>,
    id: i64,
) -> Result<u32, String> {
    metadata_service::get_driver_capabilities(&state, id)
        .await
        .map_err(String::from)
}

#[tauri::command]
pub async fn list_tables_by_conn(
    state: State<'_, AppState>,
    id: i64,
    schema: Option<String>,
) -> Result<Vec<TableInfo>, String> {
    metadata_service::list_tables_by_conn(&state, id, schema)
        .await
        .map_err(String::from)
}

File: src/services/api/metadata.ts
import { invoke } from "./core";
import { COMMANDS } from "@/services/commands";
import type {
  ConnectionForm,
  RoutineType,
} from "../types";

export const metadataApi = {
  metadata: {
    listTables: (id: number, database?: string, schema?: string) =>
      invoke(COMMANDS.LIST_TABLES, {
        id,
        database,
        schema,
      }),
    listRoutines: (id: number, database?: string, schema?: string) =>
      invoke(COMMANDS.LIST_ROUTINES, {
        id,
        database,
        schema,
      }),
    getTableStructure: (id: number, schema: string, table: string) =>
      invoke(COMMANDS.GET_TABLE_STRUCTURE, { id, schema, table }),
    getTableDDL: (
      id: number,
      database: string | undefined,
      schema: string,
      table: string,
    ) => invoke(COMMANDS.GET_TABLE_DDL, { id, database, schema, table }),
    getRoutineDDL: (
      id: number,
      database: string | undefined,
      schema: string,
      name: string,
      routineType: RoutineType,
    ) =>
      invoke(COMMANDS.GET_ROUTINE_DDL, {
        id,
        database,
        schema,
        name,
        routineType,
      }),
    getTableMetadata: (
      id: number,
      database: string | undefined,
      schema: string,
      table: string,
    ) =>
      invoke(COMMANDS.GET_TABLE_METADATA, {
        id,
        database,
        schema,
        table,
      }),
    listTablesByConn: (form: ConnectionForm) =>
      invoke(COMMANDS.LIST_TABLES_BY_CONN, { form }),
    listDatabases: (form: ConnectionForm) =>
      invoke(COMMANDS.LIST_DATABASES, { form }),
    listDatabasesById: (id: number) =>
      invoke(COMMANDS.LIST_DATABASES_BY_ID, { id }),
    getSchemaOverview: (id: number, database?: string, schema?: string) =>
      invoke(COMMANDS.GET_SCHEMA_OVERVIEW, { id, database, schema }),
    getSchemaForeignKeys: (id: number, database?: string, schema?: string) =>
      invoke(COMMANDS.GET_SCHEMA_FOREIGN_KEYS, {
        id,
        database,
        schema,
      }),
    listEvents: (id: number, database?: string, schema?: string) =>
      invoke(COMMANDS.LIST_EVENTS, { id, database, schema }),
    listSequences: (id: number, database?: string, schema?: string) =>
      invoke(COMMANDS.LIST_SEQUENCES, { id, database, schema }),
    listTypes: (id: number, database?: string, schema?: string) =>
      invoke(COMMANDS.LIST_TYPES, { id, database, schema }),
    listSynonyms: (id: number, database?: string, schema?: string) =>
      invoke(COMMANDS.LIST_SYNONYMS, { id, database, schema }),
    listPackages: (id: number, database?: string, schema?: string) =>
      invoke(COMMANDS.LIST_PACKAGES, { id, database, schema }),
    getCapabilities: (connectionId: number) =>
      invoke(COMMANDS.GET_DRIVER_CAPABILITIES, { id: connectionId }),
  },
  tableData: {
    get: (params: {
      id: number;
      database?: string;
      schema: string;
      table: string;
      page: number;
      limit: number;
      filter?: string;
      sortColumn?: string;
      sortDirection?: "asc" | "desc";
      orderBy?: string;
      includeTotal?: boolean;
    }) => invoke(COMMANDS.GET_TABLE_DATA, params),
    getByConn: (
      form: ConnectionForm,
      schema: string,
      table: string,
      page: number,
      limit: number,
      includeTotal?: boolean,
    ) =>
      invoke(COMMANDS.GET_TABLE_DATA_BY_CONN, {
        form,
        schema,
        table,
        page,
        limit,
        includeTotal,
      }),
  },
};
